cloud-sql/mysql/client-side-encryption/tests/test_encrypt_and_insert_data.py

```diff
@@ -0,0 +1,88 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import uuid
+
+import pytest
+
+from snippets.cloud_kms_env_aead import init_tink_env_aead
+from snippets.cloud_sql_connection_pool import init_db
+from snippets.encrypt_and_insert_data import encrypt_and_insert_data
+
+
+REQUIRED_ENV_VARS = [
+    "MYSQL_USER",
+    "MYSQL_PASSWORD",
+    "MYSQL_DATABASE",
+    "MYSQL_HOST",
+    "CLOUD_KMS_KEY"]
+
+table_name = f"votes_{uuid.uuid4().hex}"
+
+
+@pytest.fixture
+def check_env_vars():
+    for var in REQUIRED_ENV_VARS:
+        if not os.environ.get(var):
+            raise Exception(
+                f"Environment variable {var} must be set to perform tests.")
+    yield
+
+
+@pytest.fixture(name="pool")
+def setup_pool():
+    pool = init_db(
+        db_user=os.environ["MYSQL_USER"],
+        db_pass=os.environ["MYSQL_PASSWORD"],
+        db_name=os.environ["MYSQL_DATABASE"],
+        table_name=table_name,
+        db_host=os.environ["MYSQL_HOST"],
+
+    )
+
+    yield pool
+
+    with pool.connect() as conn:
+        conn.execute(f"DROP TABLE IF EXISTS `{table_name}`")
+
+
+@pytest.fixture(name="env_aead")
+def setup_key():
+    credentials = os.environ.get("GOOGLE_APPLICATION_CREDENTIALS", "")
+    key_uri = "gcp-kms://" + os.environ["CLOUD_KMS_KEY"]
+
+    env_aead = init_tink_env_aead(key_uri, credentials)
+
+    yield env_aead
+
+
+def test_encrypt_and_insert_data(capsys, check_env_vars, pool, env_aead):
+    encrypt_and_insert_data(
+        pool, env_aead, table_name, "SPACES", "hello@example.com")
+    captured = capsys.readouterr()
+    assert "Vote successfully cast for 'SPACES'" in captured.out
+
+    decrypted_emails = []
+    with pool.connect() as conn:
+        results = conn.execute(
+            f"SELECT team, time_cast, voter_email FROM {table_name}"
+            " ORDER BY time_cast DESC LIMIT 5").fetchall()
+
+        for row in results:
+            team = row[0]
+            email = env_aead.decrypt(row[2], team.encode()).decode()
+            decrypted_emails.append(email)
+
+        assert "hello@example.com" in decrypted_emails
```

Comment:
it seems like there's two assertions - are you testing two different things? Ideally the test should test only one thing. If it's because the actual function does both encrypting and inserting, can the encryption and insertion be split up into two functions?

Reply:
This test isn't testing two different things, it is testing one thing two different ways. The first is a quick assert to ensure that function thinks it executed correctly. The second assert is verifying in the backend that it actually did.